### DISCUSSION TITLE
What dependencies are required for `useCallback(FloatingFocusManager)`

### DISCUSSION THREAD

**Discussion by mattbrazza** [floating-ui/floating-ui#2440] [React DOM]
Hello, _TL;DR_:
Is there a way to use `useFloating().context` as a dependency within a `React.useCallback()` such that:
```tsx
const FocusWrapper = useCallback(
  ({ children }) => (<FloatingFocusManager context={context}>{children}</FloatingFocusManager>),
  [/* WANT TO PLACE `context` HERE */]
);
```

---

## Background
We are using `react v18.2.0` and `@floating-ui/react v0.24.7`.
We are trying to create hooks/components that wrap some of the Floating-UI boilerplate and help keep options consistent across our team's usage. Greatly oversimplifying things, we have something like the below working:
```tsx
function useOurFloating() {
  const { refs, floatingStyles, context, isPositioned } = useFloating({ ... });
  const click = useClick(context);
  const { getReferenceProps, getFloatingProps } = useInteractions([click]);

  const TriggerWrapper = useCallback(({ children }) => {
      return (<button ref={refs.setReference} {...getReferenceProps()}>
          {children}
        </button>);
    }, [isPositioned]);

  const PopupWrapper = useCallback(({ children }) => {
      return ReactDOM.createPortal(
          <div ref={refs.setFloating} {...getFloatingProps()} style={floatingStyles}>
            {children}
          </div>,
          document.body
        );
    }, [isPositioned]);

  return { TriggerWrapper, PopupWrapper,}
}
```

## Problem
We want to add the `<FloatingFocusManager />` to the above, but the problem is that when we do, it does not work correctly. The focus is not moved to within the Floating element and the focus is not trapped within. From what we can see it is stemming from placing the Focus Manager within the `useCallback()`; it looks something like:
```tsx
const PopupWrapper = useCallback(({ children }) => {
    return ReactDOM.createPortal(
      <FloatingFocusManager context={context}>
        <div ref={refs.setFloating} {...getFloatingProps()} style={floatingStyles}>
          {children}
        </div>
      </FloatingFocusManager>,
      document.body
    );
  }, [isPositioned]);
```

What may be an obvious thing is to add `context` to the dependency array, however since it is a complex object, the wrapper is recreated each cycle. A possible solution to that is usually to `JSON.stringify()` the object to make it a primitive, but that is not possible here due to circular referencing (stemming from `refs` and `elements`).

---

## Sample Code
[Here is a CodeSandbox URL](https://codesandbox.io/s/floating-ui-react-dom-template-forked-f9npdd?file=/src/App.js).
<details><summary>The sandbox code pasted below:</summary>
<pre><code>
import { useFloating, useInteractions, useClick, FloatingFocusManager } from "@floating-ui/react";
import { useCallback, useState } from "react";
import { createPortal } from "react-dom";

export default function App() {
  return (<div style={{ display: "flex", gap: "48px", justifyContent: "space-evenly" }}>
    <GoodCode />
    <BadCode />
  </div>);
}

function GoodCode() {
  const [isShowing, setIsShowing] = useState(false);
  const { refs, floatingStyles, context } = useFloating({ open: isShowing, onOpenChange: setIsShowing });
  const click = useClick(context);
  const { getReferenceProps, getFloatingProps } = useInteractions([click]);

  return (
    <div className="App">
      <button ref={refs.setReference} {...getReferenceProps()}>My Good Example</button>
      {isShowing &&
        createPortal(
          <FloatingFocusManager context={context}>
            <div
              ref={refs.setFloating}
              {...getFloatingProps({ style: { width: "200px", display: "flex", flexDirection: "column", ...floatingStyles } })}
              className="Tooltip"
            >
              My tooltip
              <button>Tabbable Item 1</button>
              <button>Tabbable Item 2</button>
              <button>Tabbable Item 3</button>
            </div>
          </FloatingFocusManager>,
          document.body
        )}
    </div>
  );
}

function BadCode() {
  const [isShowing, setIsShowing] = useState(false);
  const { refs, floatingStyles, context } = useFloating({ open: isShowing, onOpenChange: setIsShowing });
  const click = useClick(context);
  const { getReferenceProps, getFloatingProps } = useInteractions([click]);

  const FocusWrapper = useCallback(
    ({ children }) => (<FloatingFocusManager context={context}>{children}</FloatingFocusManager>),
    [isShowing, isPositioned]
  );

  return (
    <div className="App">
      <button ref={refs.setReference} {...getReferenceProps()}>My Good Example</button>
      {isShowing &&
        createPortal(
          <FocusWrapper>
            <div
              ref={refs.setFloating}
              {...getFloatingProps({ style: { width: "200px", display: "flex", flexDirection: "column", ...floatingStyles } })}
              className="Tooltip"
            >
              My tooltip
              <button>Tabbable Item 1</button>
              <button>Tabbable Item 2</button>
              <button>Tabbable Item 3</button>
            </div>
          </FocusWrapper>,
          document.body
        )}
    </div>
  );
}
</code></pre>
</details>

**Comment by atomiks** [comment]
I don't see a reason to `useCallback` on it at all - just return it without memoization. `floatingStyles` can also change frequently, which would also need to be included, so memoization is not effective here. The reference is also missing some deps that could cause bugs.

I would recommend reusable components instead of JSX inside `useCallback` - the floating element and context should re-render very frequently, but children inside it may be memoized if they're expensive.

**Reply by mattbrazza** [reply]
Yeah, I would like to avoid the `useCallback()`, but if without it we hit a "Maximum update depth exceeded" error:
```tsx
function useOurFloating() {
  const { refs, floatingStyles, context, isPositioned } = useFloating({ ... });

  function PopupWrapper({ children }) {
    return ReactDOM.createPortal(
      <FloatingFocusManager context={context}>
        <div ref={refs.setFloating} {...getFloatingProps()} style={floatingStyles}>
          {children}
        </div>
      </FloatingFocusManager>,
      document.body
    );
  }

  return { PopupWrapper }
}
```


The goal is to prevent an implementer from needing to deal with the boilerplate/wiring of the Floating-UI parts and just do the following:
```tsx
function SomeFeature() {
  const { TriggerWrapper, PortalWrapper } = useOurFloating(someSimpleParams);

  return (<>
    <TriggerWrapper>
      <button>The Trigger</button>
    </TriggerWrapper>
    
    <PortalWrapper>
      <SomeContent />
    </PortalWrapper>
  </>);
}
```

**Note**: The above code is all simplified and avoiding showing the `middleware`, `useInteractions()`, etc..

---

Maybe I am too far down a path and missing an obvious pitfall, which would be a relief.

**Reply by atomiks** [reply]
React doesn't support components nested inside hooks or other components. Instead define it outside the hook and pass the useFloating data to the component as prop(s). Or just specify all the deps as necessary with the `useCallback` (presumably that one doesn't cause the error?).

**Reply by mattbrazza** [reply]
> Or just specify all the deps as necessary with the useCallback (presumably that one doesn't cause the error?).

If I understand correctly, you mean to do something like `useCallback(() => <... />, [isShowing, ..., context])`? I had tried and it has the same outcome of a "max. update" error (I think stemming from the fact that dependency arrays do simple comparisons and thus fail with objects and such).

<br />

> [...] pass the useFloating data to the component as prop(s).

Yeah, that is what I would like to avoid. It would be nice for the implementers to not need to pipe through props that they should not even touch anyway. That being said, there is a reason libraries (such as Floating-UI) do it the way they do, I guess.
This may have all been wishful thinking and I will need to separate the components from the props.
